core/rawdb/accessors_trie.go

```diff
@@ -198,54 +161,51 @@ func HasTrieNode(db ethdb.KeyValueReader, owner common.Hash, path []byte, hash c
 	case HashScheme:
 		return HasLegacyTrieNode(db, hash)
 	case PathScheme:
+		var blob []byte
 		if owner == (common.Hash{}) {
-			return HasAccountTrieNode(db, path, hash)
+			blob = ReadAccountTrieNode(db, path)
+		} else {
+			blob = ReadStorageTrieNode(db, owner, path)
+		}
+		if len(blob) == 0 {
+			return false
 		}
-		return HasStorageTrieNode(db, owner, path, hash)
+		h := newHasher()
+		defer h.release()
+		return h.hash(blob) == hash // exists but not match
 	default:
 		panic(fmt.Sprintf("Unknown scheme %v", scheme))
 	}
 }
 
 // ReadTrieNode retrieves the trie node from database with the provided node info
 // and associated node hash.
-// hashScheme-based lookup requires the following:
-//   - hash
-//
-// pathScheme-based lookup requires the following:
-//   - owner
-//   - path
 func ReadTrieNode(db ethdb.KeyValueReader, owner common.Hash, path []byte, hash common.Hash, scheme string) []byte {
 	switch scheme {
 	case HashScheme:
 		return ReadLegacyTrieNode(db, hash)
 	case PathScheme:
-		var (
-			blob  []byte
-			nHash common.Hash
-		)
+		var blob []byte
 		if owner == (common.Hash{}) {
-			blob, nHash = ReadAccountTrieNode(db, path)
+			blob = ReadAccountTrieNode(db, path)
 		} else {
-			blob, nHash = ReadStorageTrieNode(db, owner, path)
+			blob = ReadStorageTrieNode(db, owner, path)
 		}
-		if nHash != hash {
-			return nil
+		h := newHasher()
+		defer h.release()
+		if h.hash(blob) != hash {
```

Comment:
`blob` can be `nil` here, in case we don't have it. If so, it's better to avoid the picking a hasher from pool and invoking the `hash` function on it. 
```suggestion
		if blob == nil {
			return nil
		}
		h := newHasher()
		defer h.release()
		if h.hash(blob) != hash {
```
Question: is it worth checking if `len(blob) == 0`? If that happens a lot (probably does not?)...